timelock-api/src/main/conjure/cross-client-batched-timelock-api.yml

```diff
@@ -0,0 +1,63 @@
+types:
+  imports:
+    LeaderTime:
+      base-type: any
+      external:
+        java: com.palantir.lock.v2.LeaderTime
+    Long:
+      base-type: any
+      external:
+        java: java.lang.Long
+    LockWatchStateUpdate:
+      base-type: any
+      external:
+        java: com.palantir.lock.watch.LockWatchStateUpdate
+    ConjureIdentifiedVersion:
+      base-type: any
+      external:
+        java: com.palantir.atlasdb.timelock.api.ConjureIdentifiedVersion
+    GetCommitTimestampsRequest:
+      base-type: any
+      external:
+        java: com.palantir.atlasdb.timelock.api.GetCommitTimestampsRequest
+    GetCommitTimestampsResponse:
+      base-type: any
+      external:
+        java: com.palantir.atlasdb.timelock.api.GetCommitTimestampsResponse
+  definitions:
+    default-package: com.palantir.atlasdb.timelock.api
+    objects:
+      NamespacedLeaderTime:
+        fields:
+          namespace: string
+          leaderTime: LeaderTime
+      NamespacedGetCommitTimestampsRequest:
+        fields:
+          namespace: string
+          getCommitTimestampsRequest: GetCommitTimestampsRequest
+      NamespacedGetCommitTimestampsResponse:
+        fields:
+          namespace: string
+          getCommitTimestampsResponse: GetCommitTimestampsResponse
+
+services:
+  CrossClientBatchedConjureTimelockService:
+    name: Cross Client Batched Timelock Service
+    default-auth: header
+    package: com.palantir.atlasdb.timelock.batch.api
+    base-path: /tl/batch
+    endpoints:
+      leaderTimes:
+        http: POST /lts
+        args:
+          namespaces: list<string>
+        returns: list<NamespacedLeaderTime>
```

Comment:
I'm not convinced this is the right approach. I'd have thought you would want a request type like `NamespacedLeaderTimes`, then call with a list of those. The point of that would be if you had incoming requests for leader time such as:

namespace1
namespace2
namespace3
namespace1

say they all get batched together - you'd likely want to further gather together all requests of the same namespace into their own requests, so that you can do clever things (coalesce / reduce overheads etc). Maybe I'm missing something - I've just looked at this briefly - but that seems like the right thing to do.

Reply:
The server should be defensive against multiple queries for same namespace.

Reply:
This is LeaderTime, though, which is a _coalescing_ endpoint, and so for this one Sudiksha's API is correct (notice that `namespaces` is a list of strings, and any futures for those namespaces will be answered with the correct answer from the list of `NamespacedLeaderTime`s).

Reply:
For leaderTime, I think having a set of namespaces should be okay. I am unsure if we need more than one computation for any client (even if they are making more than one query).

Resolving collision for NamespacedGetCommitTimestampsRequest might be trickier because of LockWatchVersion but, I can also not see the case where we manage to make two queries in one request with different lockWatchVersion for same client.

Reply:
I can:
Imagine we have one service using atlasdb-proxy that has multiple nodes. They might all talk to the same node of atlasdb-proxy but with different internal lockWatchVersions, and would expect different responses (and yet would all be the same namespace in Atlas).